[SPARK-20205][core] Make sure StageInfo is updated before sending event.

vanzin opened the pull request:
The DAGScheduler was sending a "stage submitted" event before it properly
updated the event's information. This meant that a listener (e.g. the
even logging listener) could record wrong information about the event.

This change sets the stage's submission time before the event is submitted,
when there are tasks to be executed in the stage.

Tested with existing unit tests.

vanzin commented:
@squito

SparkQA commented:
**[Test build #76697 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76697/testReport)** for PR 17925 at commit [`0d8f717`](https://github.com/apache/spark/commit/0d8f717ea722b2afec68988f8ec9705dd24d784a).

SparkQA commented:
**[Test build #76697 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76697/testReport)** for PR 17925 at commit [`0d8f717`](https://github.com/apache/spark/commit/0d8f717ea722b2afec68988f8ec9705dd24d784a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76697/
Test PASSed.

squito commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -1054,15 +1053,17 @@ class DAGScheduler(
 
+    // Mark the stage as submitted and send an event on the bus. If there are pending tasks, submit
+    // them, otherwise immediately mark the stage as finished, and submit any child stages.
+    stage.latestInfo.submissionTime = Some(clock.getTimeMillis())
```
why not just put this before the old listenerBus.post()  on L991?

this change also changes the behavior if there is a an exception while creating the tasks -- you no longer post a SparkListenerStageSubmitted.

I don't have any particular reason why you'd want the behavior one way or the other, but w/out an argument for actually changing the behavior, I'd rather do the minimal change.

mridulm replied:
Interesting, had not noticed the change in behavior.
Since behavior of `SparkListenerStageSubmitted` is unfortunately not documented, I agree that perhaps we should not change the semantics here (I am curious if it actually impacts in reality, but good to err on side of caution).

vanzin replied:
Actually I made a mental note to double-check this later, and neither option will really work because of this code in `JobProgressListener`:

```
          if (stageInfo.submissionTime.isEmpty) {
            // if this stage is pending, it won't complete, so mark it as "skipped":
            skippedStages += stageInfo
```

So the change in semantics I'm introducing is actually wrong, and I'll have to avoid it.

Commit:
[SPARK-20205][core] Make sure StageInfo is updated before sending event.

The DAGScheduler was sending a "stage submitted" event before it properly
updated the event's information. This meant that a listener (e.g. the
even logging listener) could record wrong information about the event.

This change sets the stage's submission time before the event is submitted,
when there are tasks to be executed in the stage.

SparkQA commented:
**[Test build #76819 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76819/testReport)** for PR 17925 at commit [`f1ca990`](https://github.com/apache/spark/commit/f1ca9905ef817e6374735110cce19b92cf1eecc5).

SparkQA commented:
**[Test build #76819 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76819/testReport)** for PR 17925 at commit [`f1ca990`](https://github.com/apache/spark/commit/f1ca9905ef817e6374735110cce19b92cf1eecc5).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76819/
Test FAILed.

vanzin commented:
Test failure is because of SPARK-20666. All core tests passed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #76947 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76947/testReport)** for PR 17925 at commit [`f1ca990`](https://github.com/apache/spark/commit/f1ca9905ef817e6374735110cce19b92cf1eecc5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76947/
Test PASSed.

vanzin commented:
Ping

vanzin commented:
Given the silence I assume no more feedback. Merging to master.

squito commented:
sorry didn't review this earlier, but in any case, lgtm
